### ISSUE TITLE
Unable to navigate to anotations in word 2016

### ISSUE THREAD

**Issue by pranavlal** [nvaccess/nvda#6297]
I have a document that has several annotations such as comments and track changes. I am opening this document in word 2016. I then do the following.
1. I hit nvda key + space bar to go into brows mode.
2. I then hit nvda key + f7 to activate the dialogue to list links.
3. Once that dialogue opens, I select the radio button for annotations.

At this point, NVDA appears to hang. There is no speech and I cannot do anything on the keyboard. I have to wait for about 20 to 30 seconds. I then get a tree view which lists headings. I am told that I am on level 0. I have then tried right arrowing when I am focused on a heading without success. I do not see any other annotations such as comments or track changes. The annotations are present in the document because I can hit the letter "a" from within brows mode and navigate to each annotation.
Note:
I have attached the document I am working on with this message.
Note:
When I try to exit from the tree view, NVDA freezes such that I have to reboot the computer.
os=windows 10 professional 64bit
[led_surfeit - DO.docx](https://github.com/nvaccess/nvda/files/432909/led_surfeit.-.DO.docx)

**Comment by dineshkaushal** [comment]
I have investigated this issue and found that the problem goes away when I copy text from this document to another document, comments listing starts working. But document size reduces from 147 KB to 90 KB. 

I also tried by removing the revision listing code from winword.py and comment listing started working. So it seems that word stops responding on issuing Range.Revisions command.

I also tried with another screen reader and found that word stops responding after issuing the list revision command for this document.

**Comment by Qchristensen** [comment]
The reason the document gets so much smaller (and I believe why the problem goes away) when the contents are copied to a new document, is that copying to a new document essentially resets track changes - IE everything pasted into the new document is one new change.

In this document initially, there are 1737 revisions / annotations (according to the revision pane) which includes tracked changes and comments.  When copied into a new document, there are 20, as all the tracked changes are gone, leaving only 20 comments.

I came across this in another document which only had 205 annotations.  When I deleted them until there was about 120, the elements list dialog still took about 20 seconds to load the annotations but it did load them.

**Comment by feerrenrut** [comment]
We should look into if there is any way to optimise this code. Perhaps we can reduce the amount of information fetched, or fetch it in batches.

**Comment by Qchristensen** [comment]
This came up in a thread on the user group recently, and one comment which I thought was worth recording.  From memory, I seem to recall this issue was caused by the routine timing out after a set time (20 seconds?) so that it wasn't unresponsive indefinitely.  Assuming that is correct, would it be possible to return the results retrieved within that time and perhaps have a note after the last one indicating they may be incomplete? (Ideally with a way to then retrieve the next 20 seconds worth of results)

**Comment by Adriani90** [comment]
@michaelDCurran, @leonardder I can reproduce this issue in both, Word 2016 and word 2010 with UIA disabled and enabled. In Word 2016, I get following error in the log:

```
Input: kb(laptop):shift+tab
IO - inputCore.logTimeSinceInput (18:29:56.838):
0.055 sec since input
IO - speech.speak (18:29:56.838):
Speaking [u'Typ:  Gruppe']
IO - inputCore.logTimeSinceInput (18:29:56.842):
0.059 sec since input
IO - speech.speak (18:29:56.842):
Speaking [u'Links  Auswahlschalter  aktiviert']
IO - inputCore.InputManager.executeGesture (18:29:57.734):
Input: kb(laptop):downArrow
IO - inputCore.logTimeSinceInput (18:30:04.513):
6.777 sec since input
IO - speech.speak (18:30:04.513):
Speaking [u'\xdcberschriften  Auswahlschalter  aktiviert']
IO - inputCore.InputManager.executeGesture (18:30:05.568):
Input: kb(laptop):downArrow
DEBUGWARNING - watchdog._watcher (18:30:15.746):
Trying to recover from freeze, core stack:
  File "nvda.pyw", line 233, in <module>
  File "core.pyc", line 519, in main
  File "wx\core.pyc", line 2134, in MainLoop
  File "wx\core.pyc", line 3240, in <lambda>
  File "browseMode.pyc", line 440, in run
  File "gui\__init__.pyc", line 1003, in Notify
  File "core.pyc", line 490, in run
  File "queueHandler.pyc", line 86, in pumpAll
  File "queueHandler.pyc", line 53, in flushQueue
  File "browseMode.pyc", line 870, in initElementType
  File "browseMode.pyc", line 91, in mergeQuickNavItemIterators
  File "UIABrowseMode.pyc", line 133, in UIATextAttributeQuicknavIterator
  File "documentBase.pyc", line 24, in makeTextInfo
  File "treeInterceptorHandler.pyc", line 157, in __init__
  File "NVDAObjects\UIA\__init__.pyc", line 316, in __init__

ERROR - queueHandler.flushQueue (18:30:28.339):
Error in func initElementType from eventQueue
Traceback (most recent call last):
  File "queueHandler.pyc", line 53, in flushQueue
  File "browseMode.pyc", line 902, in initElementType
  File "browseMode.pyc", line 922, in filter
  File "NVDAObjects\UIA\wordDocument.pyc", line 88, in label
TypeError: format() argument after ** must be a mapping, not NoneType
IO - inputCore.logTimeSinceInput (18:30:28.391):
22.823 sec since input
IO - speech.speak (18:30:28.391):
Speaking [u'Anmerkungen  Auswahlschalter  aktiviert']
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (18:30:28.410):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (18:30:28.415):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (18:30:28.437):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (18:30:28.441):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (18:30:28.457):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (18:30:28.463):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (18:30:28.480):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (18:30:28.483):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
DEBUGWARNING - NVDAObjects.IAccessible.IAccessible._get_IAccessibleRole (18:30:28.500):
accRole failed: (-2147024809, 'Falscher Parameter.', (None, None, None, 0, None))
IO - inputCore.InputManager.executeGesture (18:30:33.161):
Input: kb(laptop):tab
IO - inputCore.logTimeSinceInput (18:30:33.207):
0.045 sec since input
IO - speech.speak (18:30:33.207):
Speaking [u'Baumansicht']
IO - inputCore.InputManager.executeGesture (18:30:34.023):
Input: kb(laptop):shift+tab
IO - inputCore.logTimeSinceInput (18:30:34.069):
0.048 sec since input
IO - speech.speak (18:30:34.071):
Speaking [u'Typ:  Gruppe']
IO - inputCore.logTimeSinceInput (18:30:34.073):
0.052 sec since input
IO - speech.speak (18:30:34.073):
Speaking [u'Anmerkungen  Auswahlschalter  aktiviert']
IO - inputCore.InputManager.executeGesture (18:30:35.375):
Input: kb(laptop):tab
IO - inputCore.logTimeSinceInput (18:30:35.431):
0.056 sec since input
IO - speech.speak (18:30:35.431):
Speaking [u'Baumansicht']
IO - inputCore.InputManager.executeGesture (18:30:35.914):
Input: kb(laptop):downArrow
IO - inputCore.InputManager.executeGesture (18:30:36.473):
Input: kb(laptop):upArrow
```

**Comment by Adriani90** [comment]
As you can see, there is a major performance issue in the elemnts list and the tree view for annotaions is empty although there are annotations in the document. It could be also caused by a Macro which has apparently been recorded anytime in the past but I don't think so.

**Comment by Adriani90** [comment]
when pressing down arrow on the elements list radio button region to choose annotations, I face a 10 seconds delay and get following error (UIA for MS Word is disabled):

```
IO - inputCore.InputManager.executeGesture (12:09:04.734) - winInputHook (7744):
Input: kb(laptop):downArrow
DEBUGWARNING - watchdog._watcher (12:09:15.014) - watchdog (3036):
Trying to recover from freeze, core stack:
  File "nvda.pyw", line 215, in <module>
  File "core.pyc", line 545, in main
  File "wx\core.pyc", line 2134, in MainLoop
  File "wx\core.pyc", line 3240, in <lambda>
  File "browseMode.pyc", line 471, in run
  File "gui\__init__.pyc", line 1032, in Notify
  File "core.pyc", line 515, in run
  File "queueHandler.pyc", line 83, in pumpAll
  File "queueHandler.pyc", line 50, in flushQueue
  File "browseMode.pyc", line 961, in initElementType
  File "browseMode.pyc", line 95, in mergeQuickNavItemIterators
  File "NVDAObjects\window\winword.pyc", line 491, in iterate
  File "comtypes\client\lazybind.pyc", line 240, in __getitem__
  File "comtypes\automation.pyc", line 729, in _invoke
  File "comtypesMonkeyPatches.pyc", line 26, in __call__

DEBUGWARNING - queueHandler.flushQueue (12:09:15.064) - MainThread (10476):
Error in func initElementType from eventQueue
Traceback (most recent call last):
  File "queueHandler.pyc", line 50, in flushQueue
  File "browseMode.pyc", line 961, in initElementType
  File "browseMode.pyc", line 95, in mergeQuickNavItemIterators
  File "NVDAObjects\window\winword.pyc", line 491, in iterate
  File "comtypes\client\lazybind.pyc", line 240, in __getitem__
  File "comtypes\automation.pyc", line 729, in _invoke
  File "comtypesMonkeyPatches.pyc", line 34, in __call__
core.CallCancelled: COM call cancelled
IO - inputCore.logTimeSinceInput (12:09:15.099) - MainThread (10476):
10.365 sec since input
IO - speech.speak (12:09:15.099) - MainThread (10476):
Speaking ['Anmerkungen', 'Auswahlschalter', 'aktiviert']
```

**Comment by Adriani90** [comment]
This issue gets very annoying, also when pressing a in browse mode to navigate  to a change, NVDA needs about 10 seconds until it reports the change. The longer you are wokring with the document, the higher will be the performance lag in the document in general.

**Comment by Adriani90** [comment]
Still reproducible with NVDA 2020.3 and NVDA last alpha in MS Office 365 or 2016.

**Comment by Qchristensen** [comment]
Has anyone who could repro this previously tried in Office 365?  Just trying now with NVDA 2021.2.

Previously I could repro this in a document with 205 annotations.  I just created a document with 2600 annotations, and the elements list loads them all (in about 15 seconds).  It's the same PC, but Windows 10 and Office 365 both updated since then and currently at:

NVDA 2021.2 
 
Windows 10 (64-bit) Version: 2009, Build: 19043 
 
Office 365 (64-bit) Version: 16.0.14326.20220

**Comment by Qchristensen** [comment]
Meanwhile, I also just reported the same issue against Excel as issue #12846

**Comment by Adriani90** [comment]
I could reproduce a complete freeze of Windows with a laptop having Intel I12900HK 5 ghz processor, 32 gb DDR5 and NVIDIA GTX 3070 TI 8 GB grafic storage. I think we should put into question the principle of navigating to annotations via elements list entirely. I wonder if NVDA could fetch just the annotations that are visible on the screen rather than fetching all of them at once.
This means the screen should be scrolling while navigating the annotations in the elements list, so that the elements list gets updated everytime new annotations become visible on the screen.

**Comment by XLTechie** [comment]
I think @Adriani90's idea may be a good one, but for it to be viable I think 
there would need to be an announcement that new annotations are available, or 
some such.
